playback/src/audio_backend/pulseaudio.rs

```diff
@@ -3,82 +3,126 @@ use crate::config::AudioFormat;
-            _ => {
-                unimplemented!("PulseAudio currently does not support {:?} output", format)
-            }
+            _ => unreachable!(),
         };
 
         let ss = pulse::sample::Spec {
             format: pulse_format,
             channels: NUM_CHANNELS,
             rate: SAMPLE_RATE,
         };
-        debug_assert!(ss.is_valid());
 
-        Self {
-            s: None,
-            ss,
-            device,
-            format,
-        }
-    }
-}
-
-impl Sink for PulseAudioSink {
-    fn start(&mut self) -> io::Result<()> {
-        if self.s.is_some() {
-            return Ok(());
+        if !ss.is_valid() {
+            return Err(io::Error::new(
+                io::ErrorKind::Other,
+                PulseError::InvalidSampleSpec,
+            ));
         }
 
-        let device = self.device.as_deref();
         let result = Simple::new(
-            None,                // Use the default server.
-            APP_NAME,            // Our application's name.
-            Direction::Playback, // Direction.
-            device,              // Our device (sink) name.
-            STREAM_NAME,         // Description of our stream.
-            &self.ss,            // Our sample format.
-            None,                // Use default channel map.
-            None,                // Use default buffering attributes.
+            None,                   // Use the default server.
+            APP_NAME,               // Our application's name.
+            Direction::Playback,    // Direction.
+            self.device.as_deref(), // Our device (sink) name.
+            STREAM_NAME,            // Description of our stream.
+            &ss,                    // Our sample format.
+            None,                   // Use default channel map.
+            None,                   // Use default buffering attributes.
         );
+
         match result {
             Ok(s) => {
                 self.s = Some(s);
-                Ok(())
             }
-            Err(e) => Err(io::Error::new(
-                io::ErrorKind::ConnectionRefused,
-                e.to_string().unwrap(),
-            )),
+            Err(e) => {
+                return Err(io::Error::new(
+                    io::ErrorKind::ConnectionRefused,
+                    PulseError::ConnectionRefused(e),
+                ));
+            }
         }
+
+        Ok(())
     }
 
     fn stop(&mut self) -> io::Result<()> {
+        match &self.s {
+            Some(s) => {
+                if let Err(e) = s.drain() {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        PulseError::DrainFailure(e),
+                    ));
+                }
+            }
+            None => {
+                return Err(io::Error::new(
+                    io::ErrorKind::NotConnected,
+                    PulseError::ServerNone,
+                ));
+            }
+        }
```

Comment:
When there are really only two options -- `Some` or `None` -- I prefer the `if let Some(s) ... else ...` syntax to a `match`. The latter is more idiomatic when matching enum variants.

Reply:
I can do that.

Reply:
Kinda what you wanted but more short and sweet I think.

Reply:
Yeah that's nice too.